File: openInterpreterAgent/tools/lib/envelope.py
"""Tool I/O envelope helpers for the openInterpreterAgent.

Ploinky mounts each agent directory at /code. Helpers live agent-local so the
tool wrappers do not depend on a shared workspace tree.
"""

from __future__ import annotations

import json
import sys
from typing import Any, Dict, Optional


def read_envelope() -> Dict[str, Any]:
    """Read and normalize the Ploinky MCP tool envelope from stdin."""
    raw = sys.stdin.read().strip()
    if not raw:
        return {"tool": "", "input": {}, "metadata": {}}
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise ValueError(f"Invalid JSON envelope on stdin: {exc}") from exc

    return _normalize(parsed)


def _normalize(value: Any) -> Dict[str, Any]:
    if not isinstance(value, dict):
        return {"tool": "", "input": {}, "metadata": {}}

    tool = value.get("tool", "")
    if not isinstance(tool, str):
        tool = ""

    metadata = value.get("metadata") or {}
    if not isinstance(metadata, dict):
        metadata = {}

    tool_input = value.get("input")
    if isinstance(tool_input, dict):
        if (
            "input" in tool_input
            and isinstance(tool_input["input"], dict)
            and len(tool_input) == 1
        ):
            tool_input = tool_input["input"]
    elif isinstance(tool_input, str):
        try:
            parsed = json.loads(tool_input)
            tool_input = parsed if isinstance(parsed, dict) else {}
        except json.JSONDecodeError:
            tool_input = {}
    else:
        tool_input = {}

    return {"tool": tool, "input": tool_input, "metadata": metadata}


def write_ok(payload: Optional[Dict[str, Any]] = None) -> None:
    body: Dict[str, Any] = {"ok": True}
    if payload:
        for key, val in payload.items():
            if key != "ok":
                body[key] = val
    sys.stdout.write(json.dumps(body, separators=(",", ":")))


def write_error(message: str, extras: Optional[Dict[str, Any]] = None) -> None:
    body: Dict[str, Any] = {"ok": False, "error": str(message or "unknown error")}
    if extras:
        for key, val in extras.items():
            if key not in ("ok", "error"):
                body[key] = val
    sys.stdout.write(json.dumps(body, separators=(",", ":")))

File: browserUseAgent/tools/lib/envelope.mjs
import { Buffer } from 'node:buffer';

export async function readEnvelope() {
    return await new Promise((resolve, reject) => {
        const chunks = [];
        process.stdin.on('data', (chunk) => {
            chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
        });
        process.stdin.on('end', () => {
            const text = Buffer.concat(chunks).toString('utf8').trim();
            if (!text) {
                resolve({ tool: '', input: {}, metadata: {} });
                return;
            }
            try {
                const parsed = JSON.parse(text);
                resolve(normalizeEnvelope(parsed));
            } catch (err) {
                reject(new Error(`Invalid JSON envelope on stdin: ${err.message}`));
            }
        });
        process.stdin.on('error', reject);
    });
}

export function normalizeEnvelope(value) {
    if (!value || typeof value !== 'object') {
        return { tool: '', input: {}, metadata: {} };
    }
    const tool = typeof value.tool === 'string' ? value.tool : '';
    const metadata = value.metadata && typeof value.metadata === 'object' ? value.metadata : {};

    let input = value.input;
    if (input && typeof input === 'object' && !Array.isArray(input)) {
        if (input.input && typeof input.input === 'object' && !Array.isArray(input.input)
            && Object.keys(input).length === 1) {
            input = input.input;
        }
    } else if (typeof input === 'string') {
        try {
            const parsed = JSON.parse(input);
            input = parsed && typeof parsed === 'object' ? parsed : {};
        } catch {
            input = {};
        }
    } else {
        input = {};
    }
    return { tool, input, metadata };
}

export function writeOk(payload) {
    const body = { ok: true, ...payload };
    process.stdout.write(JSON.stringify(body));
}

export function writeError(message, extras) {
    const body = { ok: false, error: String(message || 'unknown error') };
    if (extras && typeof extras === 'object') {
        for (const [k, v] of Object.entries(extras)) {
            if (k !== 'ok' && k !== 'error') {
                body[k] = v;
            }
        }
    }
    process.stdout.write(JSON.stringify(body));
}
